src/library/equations_compiler/elim_match.cpp

```diff
@@ -526,26 +541,35 @@ struct elim_match_fn {
         return r && has_variable && has_constructor;
     }
 
-    /* Return true iff the next pattern of every equation is a value or variable,
+    /* Return true if the next pattern of every equation is a value or variable,
        and there are at least one equation where it is a variable and another where it is a
-       value. */
+       value.
+
+       We also perform a value transition if one of the next patterns is a
+       string literal.
+    */
     bool is_value_transition(problem const & P) {
         bool has_value    = false;
         bool has_variable = false;
+        bool has_string   = false;
         bool r = all_equations(P, [&](equation const & eqn) {
                 expr const & p = head(eqn.m_patterns);
                 if (is_local(p)) {
                     has_variable = true; return true;
                 } else {
                     type_context ctx = mk_type_context(eqn.m_lctx);
                     if (is_value(ctx, p)) {
-                        has_value    = true; return true;
+                        has_value    = true;
+                        if (is_string_value(p)) {
+                            has_string = true;
+                        }
+                        return true;
                     } else {
                         return false;
                     }
                 }
             });
-        if (!r || !has_value || !has_variable)
+        if (!has_string && (!r || !has_value || !has_variable))
             return false;
         type_context ctx  = mk_type_context(P);
```

Comment:
We have a similar problem with numerical literals.
Example: 
```lean
def f : nat → nat
| 1000 := 1
```
So, I replaced the condition above with 
```lean
if (!r || !has_value)
  return false;
```
This is sufficient for infinite types.
We also use `ite` for `char` literals. It is very unlikely someone will write function that covers all char literals, but it is possible. So, just to be on the safe side. I also added
```
if (!has_variable && has_finite_value)
    return false;
```